### DISCUSSION TITLE
Pocketbase with self-signed or commercial certs

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#930] [Q&A]
I always develop with https on.  It seems that PocketBase only works with Let's Encrypt auto certs or HTTP.

https://github.com/pocketbase/pocketbase/blob/099230a55218a3abaef3348e13abefd2af873f02/cmd/serve.go#L104

**Comment by benallfree** [comment]
Not sure if this helps, but check the docker stuff in this repo. I’m using self signed certs successfully with nginx as a proxy: https://github.com/benallfree/pockethost

**Comment by ganigeorgiev** [comment]
@dschissler That's correct. Currently the https option is only for issuing Let's encrypt certificate. If you want a custom certificate or more advanved network controls (eg. rate limiter, ip whitelist/blacklist, etc.) you can place PocketBase behind a reverse proxy as @benallfree suggested. You can find an example minimal nginx config in https://pocketbase.io/docs/going-to-production/#deploy-behind-reverse-proxy

**Comment by dschissler** [comment]
@ganigeorgiev Hey, what about having Pocketbase look for cert files with a specific name in the storage directory?  I was also thinking that there should be an `init` command that just creates the directory and core database files.  That allows dev setup scripts to 
more easily build the environment.

Possibly also with an option to add an admin account from the CLI before bringing up the webserver.  Well, even if in practice that isn't usable, then an optional `init` subcommand could allow setup options in the future.

**Comment by MrGoodbits** [comment]
@dschissler  I actually had a similar idea.  Although I think it would be more elegant to have a LetsEncrypt versus upload of the certs/key to use those specifically option in the admin pages, I respect @ganigeorgiev goal to keep the admin screens uncluttered.

I do have a working "proof of concept" that uses "pocketbase.key" and "pocketbase.crt" if they exist in the same directory as the pocketbase executable.

But, so far I haven't tested the fallback to LetsEncrypt if the files don't exist and the redirect from http->https is not finding the correct URL.

I will debug further this weekend and see about a pull request.

For me, although I understand why self-signed could be good for testing, my particular needs are that in my internal network I have a valid cert that I would like to use and the host is NOT publicly reachable so LetsEncrypt can't be used.  For my dev and small projects, I figure wrapping it behind a proxy/nginx is a bit much.

**Reply by MrGoodbits** [reply]
Ok, I have working code, but as this is only my second trip into writing code in "go", I'm sure it needs review and adjusting to better match @ganigeorgiev 's standards.

I'm also not sure the proper way to send a pull request, so I'll paste a "git diff" here.  (and learn the proper way for the future)
[diff.txt](https://github.com/pocketbase/pocketbase/files/9947930/diff.txt)


**Overview:**
With this code it will use ./pocketbase.crt (a certificate or certificate chain) and ./pocketbase.key (private key) for HTTPS if they exist.  I have tested with self-signed certs and valid wildcard certs.  If these files don't exist, and http port=80 and https port=443, it will attempt to use LetsEncrypt to acquire valid certs.  Let's Encrypt seems to require these ports.

**Issues:**
My "log.Println" are just hacky debugging.  I'm sure they can be done better.  

Even if you don't specify an http, it launches on 127.0.0.1:8090, same as default pocketbase.  I don't think it should.

Let's Encrypt requires http 80 and https 443, it warns and fails if you try other ports.  On linux, you must be root to start on those ports.

If your using self signed certs, browsers report errors and they are logged on server console. Sample:
2022/11/07 02:29:09 http: TLS handshake error from IP_ADDRESS_HERE:50747: remote error: tls: unknown certificate

I've noticed if you manually use a known good HTTPS link "https://pb.local:8093/_/#/settings" and go to the http equivalent "http://pb.local:8090/_/#/settings", it redirects back to https as expected except the url is missing the "/_/" prefix and you get "https://pb.local:8093/#/settings".  I have no idea what is going on there.  But, I wouldn't expect this to happen in actual use.

```diff
diff --git a/cmd/serve.go b/cmd/serve.go
index 1c94715..2b20c70 100644
--- a/cmd/serve.go
+++ b/cmd/serve.go
@@ -58,23 +58,48 @@ func NewServeCommand(app core.App, showStartBanner bool) *cobra.Command {
                        // start http server
                        // ---
                        mainAddr := httpAddr
+                       mainHost, httpPort, _ := net.SplitHostPort(mainAddr)
+                       httpsPort := ""
+
+                       LocalCert := false
+                       TLSConf := &tls.Config{}
+                       certManager := autocert.Manager{}
+
                        if httpsAddr != "" {
                                mainAddr = httpsAddr
+                               mainHost, httpsPort, _ = net.SplitHostPort(mainAddr)
+                               // Attempt to build tls.Config from local files
+                               cer, err := tls.LoadX509KeyPair("pocketbase.crt", "pocketbase.key")
+                               if err == nil {
+                                       // Using local cert and key
+                                       log.Println("# Using local pocketbase.key and pocketbase.crt for https")
+                                       LocalCert = true
+                                       TLSConf = &tls.Config{Certificates: []tls.Certificate{cer}}
+                               } else {
+                                       // failed with local cert, trying LetsEncrypt
+                                       log.Println("# Local pocketbase.key and pocketbase.crt for https not usable")
+                                       log.Println(err)
+                                       // LetsEncrypt ACME HTTP-01 challenge requires port 80
+                                       if (httpPort == "80" && httpsPort == "443") {
+                                               log.Println("# Will attempt to get cert from Let's Encrypt")
+                                               certManager = autocert.Manager{
+                                                       Prompt:     autocert.AcceptTOS,
+                                                       Cache:      autocert.DirCache(filepath.Join(app.DataDir(), ".autocert_cache")),
+                                                       HostPolicy: autocert.HostWhitelist(mainHost, "www."+mainHost),
+                                               }
+                                               TLSConf = &tls.Config{
+                                                       GetCertificate: certManager.GetCertificate,
+                                                       NextProtos:     []string{acme.ALPNProto},
+                                               }
+                                       } else {
+                                               log.Println("# HTTP port 80 and HTTPS port 443 is required by Let's Encrypt")
+                                       }
+                               }
                        }

-                       mainHost, _, _ := net.SplitHostPort(mainAddr)
-
-                       certManager := autocert.Manager{
-                               Prompt:     autocert.AcceptTOS,
-                               Cache:      autocert.DirCache(filepath.Join(app.DataDir(), ".autocert_cache")),
-                               HostPolicy: autocert.HostWhitelist(mainHost, "www."+mainHost),
-                       }

                        serverConfig := &http.Server{
-                               TLSConfig: &tls.Config{
-                                       GetCertificate: certManager.GetCertificate,
-                                       NextProtos:     []string{acme.ALPNProto},
-                               },
+                               TLSConfig: TLSConf,
                                ReadTimeout: 60 * time.Second,
                                // WriteTimeout: 60 * time.Second, // breaks sse!
                                Handler: router,
@@ -97,7 +122,11 @@ func NewServeCommand(app core.App, showStartBanner bool) *cobra.Command {
                        if httpsAddr != "" {
                                // if httpAddr is set, start an HTTP server to redirect the traffic to the HTTPS version
                                if httpAddr != "" {
-                                       go http.ListenAndServe(httpAddr, certManager.HTTPHandler(nil))
+                                       if LocalCert {
+                                               go http.ListenAndServe(httpAddr, http.RedirectHandler("https://"+serverConfig.Addr+"/", 301))
+                                       } else {
+                                               go http.ListenAndServe(httpAddr, certManager.HTTPHandler(nil))
+                                       }
                                }

                                // start HTTPS server

```

**Reply by MrGoodbits** [reply]
I realized the forcing of ports 80 and 443 to allow LetsEcnrypt to even launch might be a bit aggressive.  The public facing ports must be that for getting the certs, but if pocketbase is behind a NAT/Proxy it could be on other ports and still successfully acquire a certificate.

It is worth noting the certManager's built in redirector is hard coded to port 443 so public facing 80/443 is stuck for LetsEncrypt.

**Reply by ganigeorgiev** [reply]
@MrGoodbits I'm not sure that I understand the proposed changes. For now there are no plans to support self-signed certificates when using the `--https` flag. In those cases most likely you'll already have a reverse proxy in place and I don't think it would be a good idea to complicate the initialization setup with fixed files lookup. In the future we may separate the internals of the `serve` command into its own function and have options argument with a tls config member to allow users to customize the behavior if they want to.

**Reply by dschissler** [reply]
@ganigeorgiev In my case it's not self-signed certificates, but rather expensive ones that require a lot of verification.  It would be nice to be able to run them directly.  It's not a show stopper for me though because I can just put a reverse proxy in front.  However, all of these extra steps are things that get in the way of being able to quickly get up to speed with Pocketbase, even if it's not the final way to do it.

Just last night I ported to Echo a Chi handler that tries to use a Brotli file from an Golang embed if it exists and then tries the uncompressed file.  So I'm saying that I can do this stuff, but at this moment I really don't want to play with Caddy (which I like a lot btw) because I have more important things to dig into as I port my early app over to Pocketbase.  I'd prefer to spend those moments elsewhere, such as discovering any roadblocks as soon as possible so that it can get worked out.

I can imagine a huge amount of people who know how to install a SSL cert into Apache via command line, and are just learning Golang from PHP, and have learned that Pocketbase is something that they could start with.  These people would be better served spending their time learning Golang and Pocketbase than needing to first use another tool that is standing in their way.

**Reply by shadowempire123** [reply]
I also wanted to note that it would be nice to be able to use self signed certs. In my company I have strict firewalls rules and I dont have access to letsencrypt or root permissions.
Even setting up a reverse proxy is a pain (including several departments, approvals,... :-/)
I developed an internal tool which must have some kind of authentication, and pocketbase would work out of the box (using this tut:
https://www.youtube.com/watch?v=doDKaKDvB30)

I have a sveltekit frontend app and an node.js backend app running with selfsigned certs. Just missing an authentication GUI... :-/

**Reply by dschissler** [reply]
@shadowempire123 "Open source backend in 2 files" :)

**Reply by shadowempire123** [reply]
> @shadowempire123 "Open source backend in 2 files" :)

what do you mean? 😲

**Reply by dschissler** [reply]
@shadowempire123 It's the Pocketbase slogan.

**Comment by n1xx1** [comment]
Instead of NGINX that is really awesome for production, for local development I would suggest running [Caddy](https://caddyserver.com/) with a `Caddyfile` that looks like this:

```
example.com {
    tls internal {

    }
    reverse_proxy 127.0.0.1:8090
}
```

If run as root it will also automatically install its self signed certificate.
